hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/web/resources/NamenodeWebHdfsMethods.java

```diff
@@ -1345,11 +1348,21 @@ protected Response get(
     }
   }
 
-  private static String getTrashRoot(String fullPath,
-      Configuration conf) throws IOException {
-    FileSystem fs = FileSystem.get(conf != null ? conf : new Configuration());
-    return fs.getTrashRoot(
-        new org.apache.hadoop.fs.Path(fullPath)).toUri().getPath();
+  private String getTrashRoot(String fullPath) throws IOException {
+    String user = UserGroupInformation.getCurrentUser().getShortUserName();
+    org.apache.hadoop.fs.Path path = new org.apache.hadoop.fs.Path(fullPath);
+    String parentSrc = path.isRoot() ?
+        path.toUri().getPath() : path.getParent().toUri().getPath();
+    EncryptionZone ez = getRpcClientProtocol().getEZForPath(parentSrc);
+    org.apache.hadoop.fs.Path trashRoot;
+    if (ez != null) {
+      trashRoot = new org.apache.hadoop.fs.Path(
+          new org.apache.hadoop.fs.Path(ez.getPath(), TRASH_PREFIX), user);
+    } else {
+      trashRoot = new org.apache.hadoop.fs.Path(
+          new org.apache.hadoop.fs.Path(USER_HOME_PREFIX, user), TRASH_PREFIX);
+    }
+    return trashRoot.toUri().getPath();
```

Comment:
Thanks for the comment. I moved the common logic into DFSUtilClient in the last commit. While added `TestWebHDFS#testGetEZTrashRoot` could detect the future mismatch, it is not in the same subproject of DistributedFileSystem (hadoop-hdfs-client).

Reply:
Discussed with @daryn-sharp on this.  New features in the future might require additional logic in determining trash root. It is better to be done at the server-side.  So, the Ideal solution would be to add a `getTrashRoot()` namenode RPC method. `DistributedFileSystem` would then have a fallback logic for compatibility with older servers.  `NamenodeWebHdfsMethods` would simply call this rpc method.

Also, use of `Path` on NN is highly discouraged. Path is super-expensive and performs normalization that no other operation performs.

Reply:
From the the current implementation, we need to fix the memory leak and the fact it does not work with security enabled, and probably backport that across the active branches. The memory leak especially is a big problem.

Would it make sense to adopt this patch as it stands and push it to all the branches and then open another Jira to implement the namenode RPC suggestion on trunk as a new feature?

Reply:
I am fine with an interim solution to fix the immediate issues. But use of `Path` in NN is far from ideal. It will be best if this is addressed now than later.

Reply:
@kihwal Thanks for the comment. I updated the code based on your suggestion.